packages/malloy/src/model/malloy_types.ts

```diff
@@ -434,38 +431,106 @@ export function mkExpr(
   return ret;
 }
 
-export type ExpressionType = 'scalar' | 'aggregate' | 'analytic';
+export type ExpressionType =
+  | 'scalar'
+  | 'aggregate'
+  | 'scalar_analytic'
+  | 'aggregate_analytic'
+  | 'ungrouped_aggregate';
 
 export interface Expression {
   e?: Expr;
   expressionType?: ExpressionType;
   code?: string;
 }
 
+export function expressionIsScalar(e: ExpressionType | undefined): boolean {
+  return e === undefined || e === 'scalar';
+}
+
 export function expressionIsAggregate(e: ExpressionType | undefined): boolean {
-  return e === 'aggregate';
+  return e === 'aggregate' || e === 'ungrouped_aggregate';
+}
+
+export function expressionIsUngroupedAggregate(
+  e: ExpressionType | undefined
+): boolean {
+  return e === 'ungrouped_aggregate';
+}
+
+export function expressionInvolvesAggregate(
+  e: ExpressionType | undefined
+): boolean {
+  return (
+    e === 'aggregate' ||
+    e === 'ungrouped_aggregate' ||
+    e === 'aggregate_analytic'
+  );
 }
 
 export function expressionIsCalculation(
   e: ExpressionType | undefined
 ): boolean {
-  return e === 'aggregate' || e === 'analytic';
+  return (
+    e === 'aggregate' ||
+    e === 'scalar_analytic' ||
+    e === 'aggregate_analytic' ||
+    e === 'ungrouped_aggregate'
+  );
+}
+
+export function expressionIsAnalytic(e: ExpressionType | undefined): boolean {
+  return e === 'aggregate_analytic' || e === 'scalar_analytic';
+}
+
+function expressionTypeLevel(e: ExpressionType): number {
+  return {
+    scalar: 0,
+    aggregate: 1,
+    ungrouped_aggregate: 2,
+    scalar_analytic: 2,
+    aggregate_analytic: 3,
+  }[e];
+}
+
+export function isExpressionTypeLEQ(
+  e1: ExpressionType,
+  e2: ExpressionType
+): boolean {
+  return e1 === e2 || expressionTypeLevel(e1) < expressionTypeLevel(e2);
 }
 
+// TODO rename this to be like `combineExpressionTypes`
 export function maxExpressionType(
   e1: ExpressionType,
   e2: ExpressionType
 ): ExpressionType {
+  // TODO handle the case where e1 is analytic and e2 is ungrouped_aggregate
   let ret: ExpressionType = 'scalar';
   if (e1 === 'aggregate' || e2 === 'aggregate') {
     ret = 'aggregate';
   }
-  if (e1 === 'analytic' || e2 === 'analytic') {
-    ret = 'analytic';
+  if (e1 === 'ungrouped_aggregate' || e2 === 'ungrouped_aggregate') {
+    ret = 'ungrouped_aggregate';
+  }
+  if (e1 === 'scalar_analytic' || e2 === 'scalar_analytic') {
+    ret = 'scalar_analytic';
+  }
+  if (e1 === 'aggregate_analytic' || e2 === 'aggregate_analytic') {
+    ret = 'aggregate_analytic';
+  }
+  if (e1 === 'scalar_analytic' && e2 === 'aggregate') {
+    ret = 'aggregate_analytic';
+  } else if (e1 === 'aggregate' && e2 === 'scalar_analytic') {
+    ret = 'aggregate_analytic';
```

Comment:
I don't have the brain power right now to think about this case of `analytic` + `ungrouped_aggregate`. Maybe need to think about it.